File: bestlocationstrategy/src/main/java/com/location/bestlocationstrategy/LocationManagerStrategy.java
package com.location.bestlocationstrategy;

import android.content.Context;
import android.location.Criteria;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;

/**
 * Created by Balwinder on 20/Sept/2018.
 */

public class LocationManagerStrategy implements
        BaseLocationStrategy,
        LocationListener {

    private Context mAppContext;
    private LocationManager mLocationManager;
    private Location mLastLocation;
    private LocationChangesListener mLocationListener;
    private boolean mUpdatePeriodically = false;
    private static LocationManagerStrategy INSTANCE;
    // Location updates intervals in sec
    private static long UPDATE_INTERVAL = 10000; // 10 sec
    private static long FASTEST_INTERVAL = 5000; // 5 sec
    private static long DISPLACEMENT = 10; // 10 meters
    // flag for GPS status
    private boolean isGPSEnabled = false;
    // flag for network status
    private boolean isNetworkEnabled = false;
    private static final String TAG = "LocationManagerStrategy";



    public LocationManagerStrategy(Context context) {
        this.mAppContext = context;
    }


    public static LocationManagerStrategy getInstance(Context context) {
        if (INSTANCE == null) {
            INSTANCE = new LocationManagerStrategy(context);
            INSTANCE.initLocationClient();
        }
        return INSTANCE;
    }

    @Override
    public void startListeningForLocationChanges(LocationChangesListener locationListener) {
        mLocationListener = locationListener;
        startLocationUpdates();
    }

    @Override
    public void stopListeningForLocationChanges() {
        try {
            mLocationManager.removeUpdates(this);
        } catch (SecurityException ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void setPeriodicalUpdateEnabled(boolean enable) {
        mUpdatePeriodically = enable;
    }

    @Override
    public void setPeriodicalUpdateTime(long time) {
        UPDATE_INTERVAL = time;
    }

    @Override
    public void setDisplacement(long displacement) {
        DISPLACEMENT = displacement;
    }

    @Override
    public Location getLastLocation() {
        if (mLastLocation == null) {
            try {
                mLastLocation = mLocationManager.getLastKnownLocation(getBestProvider());
            } catch (SecurityException securityException) {
                return null;
            }
        }
        LocationUtils.LastKnownLocaiton = mLastLocation;
        return mLastLocation;
    }

    private String getBestProvider() {
        Criteria criteria = new Criteria();
        criteria.setAccuracy(Criteria.ACCURACY_HIGH);
        return mLocationManager.getBestProvider(criteria, false);
    }

    @Override
    public void initLocationClient() {
        mLocationManager = (LocationManager) mAppContext.getSystemService(Context.LOCATION_SERVICE);
    }

    @Override
    public void startLocationUpdates() {
        try {
            // getting GPS status
            isGPSEnabled = mLocationManager
                    .isProviderEnabled(LocationManager.GPS_PROVIDER);
            // getting network status
            isNetworkEnabled = mLocationManager
                    .isProviderEnabled(LocationManager.NETWORK_PROVIDER);
            if(isNetworkEnabled) {
                mLocationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,
                        UPDATE_INTERVAL,
                        DISPLACEMENT, this);
            }
            if(isGPSEnabled) {
                mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER,
                        UPDATE_INTERVAL,
                        DISPLACEMENT, this);
            }
        } catch (SecurityException ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void onLocationChanged(Location location) {
        mLastLocation = location;
        if(mLastLocation!=null && mLocationListener!=null && LocationUtils.isBetterLocation(location)) {
            mLocationListener.onLocationChanged(location);
        }
        if (!mUpdatePeriodically) {
            this.stopListeningForLocationChanges();
        }
    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {
        if(mLocationListener!=null)
        mLocationListener.onConnectionStatusChanged();
    }

    @Override
    public void onProviderEnabled(String provider) {
        if(mLocationListener!=null)
        mLocationListener.onConnected();
    }

    @Override
    public void onProviderDisabled(String provider) {
        if(mLocationListener!=null)
        mLocationListener.onFailure(provider);
    }
}

File: bestlocationstrategy/src/main/java/com/location/bestlocationstrategy/LocationManagerStrategy.kt
package com.location.bestlocationstrategy

import android.content.Context
import android.location.Criteria
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle

/**
 * Created by Balwinder on 20/Sept/2018.
 */
class LocationManagerStrategy(private val mAppContext: Context) : BaseLocationStrategy, LocationListener {
    private var mLocationManager: LocationManager? = null
    private var mLastLocation: Location? = null
    private var mLocationListener: LocationChangesListener? = null
    private var mUpdatePeriodically = false

    // flag for GPS status
    private var isGPSEnabled = false

    // flag for network status
    private var isNetworkEnabled = false
    override fun startListeningForLocationChanges(locationListener: LocationChangesListener?) {
        mLocationListener = locationListener
        startLocationUpdates()
    }

    override fun stopListeningForLocationChanges() {
        try {
            mLocationManager!!.removeUpdates(this)
        } catch (ex: SecurityException) {
            ex.printStackTrace()
        }
    }

    override fun setPeriodicalUpdateEnabled(enable: Boolean) {
        mUpdatePeriodically = enable
    }

    override fun setPeriodicalUpdateTime(time: Long) {
        UPDATE_INTERVAL = time
    }

    override fun setDisplacement(displacement: Long) {
        DISPLACEMENT = displacement
    }

    override val getLatestLocation: Location?
        get() {
            if (mLastLocation == null) {
                mLastLocation = try {
                    mLocationManager!!.getLastKnownLocation(bestProvider)
                } catch (securityException: SecurityException) {
                    return null
                }
            }
            return mLastLocation
        }

    private val bestProvider: String
        get() {
            val criteria = Criteria()
            criteria.accuracy = Criteria.ACCURACY_HIGH
            return mLocationManager!!.getBestProvider(criteria, false)!!
        }

    override fun initLocationClient() {
        mLocationManager = mAppContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
    }

    override fun startLocationUpdates() {
        try {
            // getting GPS status
            isGPSEnabled = mLocationManager?.isProviderEnabled(LocationManager.GPS_PROVIDER)!!
            // getting network status
            isNetworkEnabled = mLocationManager?.isProviderEnabled(LocationManager.NETWORK_PROVIDER)!!


            if (isNetworkEnabled) {
                mLocationManager!!.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,
                        UPDATE_INTERVAL,
                        DISPLACEMENT.toFloat(), this)
            }

            if (isGPSEnabled) {
                mLocationManager!!.requestLocationUpdates(LocationManager.GPS_PROVIDER,
                        UPDATE_INTERVAL,
                        DISPLACEMENT.toFloat(), this)
            }

        } catch (ex: SecurityException) {
            ex.printStackTrace()
        }
    }

    override fun fetchLocationInBackgroundBelowQ(enable: Boolean) {
     //   TODO("Not yet implemented")
    }

    override fun getBackGroundLocationQandAbove() {
       // TODO("Not yet implemented")
    }

    override fun onLocationChanged(location: Location) {
        mLastLocation = location
        if (mLastLocation != null && mLocationListener != null && LocationUtils.isBetterLocation(location)) {
            mLocationListener!!.onBetterLocationAvailable(location)
        }
        if (!mUpdatePeriodically) {
            stopListeningForLocationChanges()
        }
    }

    override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {
        if (mLocationListener != null) mLocationListener!!.onConnectionStatusChanged()
    }

    override fun onProviderEnabled(provider: String) {
        if (mLocationListener != null) mLocationListener!!.onConnected()
    }

    override fun onProviderDisabled(provider: String) {
        if (mLocationListener != null) mLocationListener!!.onFailure(provider)
    }

    companion object {
        private var INSTANCE: LocationManagerStrategy? = null

        // Location updates intervals in sec
        private var UPDATE_INTERVAL: Long = 10000 // 10 sec
        private const val FASTEST_INTERVAL: Long = 5000 // 5 sec
        private var DISPLACEMENT: Long = 10 // 10 meters
        private const val TAG = "LocationManagerStrategy"
        @JvmStatic
        fun getInstance(context: Context): LocationManagerStrategy? {
            if (INSTANCE == null) {
                INSTANCE = LocationManagerStrategy(context)
                INSTANCE!!.initLocationClient()
            }
            return INSTANCE
        }
    }

}